### PR TITLE
Documentation Update for 0.3.2

### PR DESCRIPTION
**Pull request by Spartee**
The readme has been updated with new examples and
usage patterns.

A new Online Analysis examples has been created with
a Lattice Boltzman simulation that will show users
how to perform streaming analysis with Smartsim.

### CODE REVIEW

**Review by al-rigazzi — CHANGES_REQUESTED** [review]
Just some typos here and there. Looks stellar.

File: README.md [hunk 1]
```diff
@@ -90,43 +457,513 @@ API with additions specific to scientific workflows.
...
+    ],
+    name="FCN",
+)
+
+# Compile model with optimizer
+model.compile(optimizer="adam",
+             loss="sparse_categorical_crossentropy",
+             metrics=["accuracy"])
+
+# training code ommited
```
Comment (EricGustin) [inline README.md:849]: omitted

File: README.md [hunk 2]
```diff
@@ -90,43 +457,513 @@ API with additions specific to scientific workflows.
...
+models to ONNX.
+
+ - [onnxmltools](https://github.com/onnx/onnxmltools)
+ - [skl2onnx](https://github.com/onnx/sklearn-onnx/)
+ - [tensorflow-onnx](https://github.com/onnx/tensorflow-onnx/)
+
+And PyTorch has it's own converter.
+
+Below are some examples of a few models in Sci-kit Learn that are converted
+into onnx format for use with SmartSim.
```
Comment (EricGustin) [inline README.md:895]: for consistency, `onnx` -> `ONNX`

File: README.md [hunk 3]
```diff
@@ -90,43 +457,513 @@ API with additions specific to scientific workflows.
...
+#### KMeans
+
+K-means clustering is an unsupervised ML algorithm. It is used to categorize data points
+into f groups ("clusters"). Sci-kit Learn has a built in implementation of K-means clustering
```
Comment (EricGustin) [inline README.md:899]: `built in` -> `built-in`

File: README.md [hunk 4]
```diff
@@ -90,43 +457,513 @@ API with additions specific to scientific workflows.
...
+
+client.set_model("kmeans", model, "ONNX", device="CPU")
+client.run_model("kmeans", inputs="input", outputs=["labels", "transform"])
+
+print(client.get_tensor("labels"))
+```
+
+#### Random Forest
+
+The Random Forest example uses the Iris datset from Sci-kit Learn to train a
```
Comment (EricGustin) [inline README.md:922]: dataset

**Review by EricGustin — COMMENTED** [review]
Some more typos to address.

File: CITATION.cff [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  - family-names: Ellis
+    given-names: Matthew
+    orcid: https://orcid.org/0000-0002-5782-5447
+  - family-names: Rigazzi
+    given-names: Alessandro
+    orcid: https://orcid.org/0000-0003-2132-7726
+  - family-names: Bachman
+    given-names: Scott
+    orcid: https://orcid.org/0000-0002-6479-4300
+  - family-names: Gustavo
```
Comment (EricGustin) [inline CITATION.cff:16]:
```suggestion
  - family-names: Marques
```

File: CITATION.cff [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    given-names: Marques
```
Comment (EricGustin) [inline CITATION.cff:17]:
```suggestion
    given-names: Gustavo
```

File: CITATION.cff [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    orcid: https://orcid.org/0000-0001-7238-0290
+  - family-names: Andrew
+    given-names: Shao
+    orcid: https://orcid.org/0000-0003-3658-512X
+  - family-names: Benjamin
+    given-names: Robbins
+    orcid: https://orcid.org/0000-0003-3658-512X
+
+title: "Using Machine Learning at Scale in HPCSimulations with SmartSim: An Application to Ocean Climate Modeling"
```
Comment (EricGustin) [inline CITATION.cff:26]:
```suggestion
title: "Using Machine Learning at Scale in HPC Simulations with SmartSim: An Application to Ocean Climate Modeling"
```

File: doc/tutorials/lattice_boltz_analysis.rst [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+are going to be launching the simulation through SmartSim.
+
+The full simulation code can be found here # PUT IN LINK
+
+3.2 Creating the Analysis Driver
+--------------------------------
+
+
+SmartSim, the infrastructure library, is used here to launch both the
+database and the simulation locally, but in seperate processes. The example
```
Comment (EricGustin) [inline doc/tutorials/lattice_boltz_analysis.rst:80]:
```suggestion
database and the simulation locally, but in separate processes. The example
```

File: doc/tutorials/lattice_boltz_analysis.rst [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+is designed to run on laptops, so the local launcher is used.
+
+
+First the necessary libraries are imported and and `Experiment` instance is created.
+Also defined is an `Orchestrator` which is the refrence to the in-memory database
```
Comment (EricGustin) [inline doc/tutorials/lattice_boltz_analysis.rst:85]:
```suggestion
Also defined is an `Orchestrator` which is the reference to the in-memory database
```

File: doc/tutorials/lattice_boltz_analysis.rst [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+The simulation is started with `block=False`, so that the data
+being streamed from the simulation can be analyzed in real time.
+
+.. code-block:: python
+
+    # start the database and connect client to get data
+    exp.start(db)
+    client = Client(address="127.0.0.1:6780", cluster=False)
+
+    # start simulation without blocking so data can be analyized in real time
```
Comment (EricGustin) [inline doc/tutorials/lattice_boltz_analysis.rst:139]:
```suggestion
    # start simulation without blocking so data can be analyzed in real time
```

File: doc/tutorials/lattice_boltz_analysis.rst [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+    exp.start(model, block=False, summary=True)
+
+
+SmartRedis is used  to pull the Datasets created by
```
Comment (EricGustin) [inline doc/tutorials/lattice_boltz_analysis.rst:143]:
```suggestion
SmartRedis is used to pull the Datasets created by
```

File: doc/tutorials/lattice_boltz_analysis.rst [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+the simulation and use matplotlib to plot the results.
+
+Another `Model` could have been created to plot the results and launched
+in a similar manner to the simulation.
+
+Doing so would enable the analysis application to be executed on different
+resources such as GPU enabled nodes, or distributed accross nodes.
```
Comment (EricGustin) [inline doc/tutorials/lattice_boltz_analysis.rst:150]:
```suggestion
resources such as GPU enabled nodes, or distributed across nodes.
```

File: doc/tutorials/lattice_boltz_analysis.rst [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+
+This version, where the driver and anaylsis code coexist in the same
```
Comment (EricGustin) [inline doc/tutorials/lattice_boltz_analysis.rst:152]:
```suggestion
This version, where the driver and analysis code coexist in the same
```

File: tutorials/03_online_analysis/lattice/driver.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+model.attach_generator_files(to_copy="fv_sim.py")
+
+# generate directories for output, error and results
+exp.generate(db, model, overwrite=True)
+
+# start the database and connect client to get data
+exp.start(db)
+client = Client(address="127.0.0.1:6780", cluster=False)
+
+# start simulation without blocking so data can be analyized in real time
```
Comment (EricGustin) [inline tutorials/03_online_analysis/lattice/driver.py:33]:
```suggestion
# start simulation without blocking so data can be analyzed in real time
```

File: README.md [hunk 15]
```diff
@@ -32,56 +32,419 @@
...
+SmartSim provides an API to connect HPC workloads, particularly (MPI + X) simulations,
+to an in-memory database called the Orchestrator, built on an in-memory database
+called Redis.
 
-In addition, the Orchestrator is equipped with ML inference runtimes:
-PyTorch, TensorFlow, and ONNX. From inside a simulation, users can
-store and execute trained models and retrieve the result.
+Applications integrated with the SmartRedis clients, written in Fortran, C, C++ and Python,
+can stream tensors and datasets to and from the Orchestrator This Client-Server
+paradigm allows for data to be seemlessly exchanged between applications at runtime.
```
Comment (EricGustin) [inline README.md:43]:
```suggestion
paradigm allows for data to be seamlessly exchanged between applications at runtime.
```

File: README.md [hunk 16]
```diff
@@ -32,56 +32,419 @@
...
 
+In addition to exchanging data between langauges, any of the SmartRedis clients can
```
Comment (EricGustin) [inline README.md:45]:
```suggestion
In addition to exchanging data between languages, any of the SmartRedis clients can
```

File: README.md [hunk 17]
```diff
@@ -32,56 +32,419 @@
...
+remotely execute Machine Learning models and TorchScript code on data stored in
+the Orchestrator dispite which langauge the data originated from.
```
Comment (EricGustin) [inline README.md:47]:
```suggestion
the Orchestrator despite which language the data originated from.
```

File: README.md [hunk 18]
```diff
@@ -32,56 +32,419 @@
...
+```
+```bash
+# on Slurm system
+sbatch run_hello_world.sh
+```
+
+### Batch Launch Examples
+
+SmartSim can also launch workloads in a batch directly from Python, without the need
+for a batch script. Users can launch groups of ``Model`` instances in a ``Ensemble``.
```
Comment (EricGustin) [inline README.md:226]:
```suggestion
for a batch script. Users can launch groups of ``Model`` instances in an ``Ensemble``.
```

File: README.md [hunk 19]
```diff
@@ -32,56 +32,419 @@
...
+
+The following launches 4 replicas of the the same ``hello_world`` model.
```
Comment (EricGustin) [inline README.md:228]:
```suggestion
The following launches 4 replicas of the same ``hello_world`` model.
```

File: README.md [hunk 20]
```diff
@@ -32,56 +32,419 @@
...
+ - Orchestrator - In-memory data store and Machine Learning Inference (Redis + RedisAI)
+ - Ray - Distributed Reinforcement Learning (RL), Hyperparameter Optimization (HPO)
+
+## Redis + RedisAI
+
+The ``Orchestrator`` is an in-memory database that utilizes Redis and RedisAI to provide
+a distributed database and access to ML runtimes from Fortran, C, C++ and Python.
+
+SmartSim provides classes that make it simple to launch the database in many
+configurations and optional form a distributed database cluster. The examples
```
Comment (EricGustin) [inline README.md:302]:
```suggestion
configurations and optionally form a distributed database cluster. The examples
```

File: README.md [hunk 21]
```diff
@@ -32,56 +32,419 @@
...
+from smartsim import Experiment
+from smartsim.database import Orchestrator
+
+exp = Experiment("local-db", launcher="local")
+db = Orchestrator(port=6780)
+
+# by default, SmartSim never blocks execution after the database is launched.
+exp.start(db)
+
+# launch models, anaylsis, training, inference sessions, etc
```
Comment (EricGustin) [inline README.md:322]:
```suggestion
# launch models, analysis, training, inference sessions, etc
```

File: README.md [hunk 22]
```diff
@@ -32,56 +32,419 @@
...
+db_cluster = SlurmOrchestrator(db_nodes=3, db_port=6780, batch=False)
 
-The following third-party (non-Python) libraries are used in the SmartSim IL.
+exp.start(db_cluster)
 
- - [Redis](https://github.com/redis/redis) 6.0.8
- - [RedisAI](https://github.com/RedisAI/RedisAI) 1.2
- - [RedisIP](https://github.com/Spartee/RedisIP) 0.1.0
+print(f"Orchestrator launched on nodes: {db_cluster.hosts}")
+# launch models, anaylsis, training, inference sessions, etc
```
Comment (EricGustin) [inline README.md:356]:
```suggestion
# launch models, analysis, training, inference sessions, etc
```

File: README.md [hunk 23]
```diff
@@ -32,56 +32,419 @@
...
+from smartsim import Experiment
+from smartsim.database import PBSOrchestrator
+
+exp = Experiment("db-on-slurm", launcher="slurm")
+db_cluster = PBSOrchestrator(db_nodes=3, db_port=6780, batch=False)
+
+exp.start(db_cluster)
+
+print(f"Orchestrator launched on nodes: {db_cluster.hosts}")
+# launch models, anaylsis, training, inference sessions, etc
```
Comment (EricGustin) [inline README.md:383]:
```suggestion
# launch models, analysis, training, inference sessions, etc
```

File: README.md [hunk 24]
```diff
@@ -32,56 +32,419 @@
...
+from smartsim.database import PBSOrchestrator
+
+exp = Experiment("db-on-slurm", launcher="pbs")
+db_cluster = PBSOrchestrator(db_nodes=3, db_port=6780, batch=True,
+                             time="00:10:00", account="12345-Cray", queue="cl40")
+
+exp.start(db_cluster)
+
+print(f"Orchestrator launched on nodes: {db_cluster.hosts}")
+# launch models, anaylsis, training, inference sessions, etc
```
Comment (EricGustin) [inline README.md:411]:
```suggestion
# launch models, analysis, training, inference sessions, etc
```

File: README.md [hunk 25]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
-|----------	|:----------------:	|
-| Python   	|        3.7+      	|
-| Fortran  	|       2003       	|
-| C        	|        C99       	|
-| C++      	|       C++11      	|
+## Tensors
+
+Tensors are the fundamental data structure for the SmartRedis clients. The Clients
+use the native array format of the language. For example, in Python, a tensor is
+a NumPy array. The C++/C client accepts nested and contingous arrays.
```
Comment (EricGustin) [inline README.md:460]:
```suggestion
a NumPy array. The C++/C client accepts nested and contiguous arrays.
```

File: README.md [hunk 26]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+
+When stored in the database, all tensors are stored in the same format. Hence,
+any language can recieve a tensor from the database no matter what supported language
```
Comment (EricGustin) [inline README.md:463]:
```suggestion
any language can receive a tensor from the database no matter what supported language
```

File: README.md [hunk 27]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+
+For an example of how to use the ``Dataset`` class, see the [Online Analysis example](#online-analysis)
+
+For more information on the API, see the
+[API documentation](https://www.craylabs.org/docs/sr_data_structures.html#dataset)
+
+## Examples
+
+Even though the clients rely on the Orchestrator database to be running, it can be helpful
+to see examples of how the API is used accross different languages even without the
```
Comment (EricGustin) [inline README.md:484]:
```suggestion
to see examples of how the API is used across different languages even without the
```

File: README.md [hunk 28]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+infrastructure code. The following examples provide simple examples of client usage.
+
+For more imformation on the SmartRedis clients, see the
```
Comment (EricGustin) [inline README.md:487]:
```suggestion
For more information on the SmartRedis clients, see the
```

File: README.md [hunk 29]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+[API documentation](https://www.craylabs.org/docs/api/smartredis_api.html) and
+[tutorials](https://www.craylabs.org/docs/tutorials/smartredis.html).
+
+**Please note** these are client examples, they will not run if there is no database to
+connect to.
+
+### Python
+
+Training code and Model construction are not shown here, but the example below
+shows how to take a PyTorch model, sent it to the database, and execute it
```
Comment (EricGustin) [inline README.md:497]:
```suggestion
shows how to take a PyTorch model, send it to the database, and execute it
```

File: README.md [hunk 30]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+  real, dimension(1,1,mnist_dim1,mnist_dim2) :: array
+  real, dimension(1,result_dim1) :: result
+
+  character(len=255) :: in_key
+  character(len=255) :: out_key
+
+  character(len=255), dimension(1) :: inputs
+  character(len=255), dimension(1) :: outputs
+
+  ! Construct the keys used for the specifiying inputs and outputs
```
Comment (EricGustin) [inline README.md:608]:
```suggestion
  ! Construct the keys used for the specifying inputs and outputs
```

File: README.md [hunk 31]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+end program run_mnist_example
+
+```
+
+
+---------
+# SmartSim + SmartRedis
+
+SmartSim and SmartRedis were designed to work together. When launched through
+SmartSim, applcations using the SmartRedis clients are directly connected to
```
Comment (EricGustin) [inline README.md:633]:
```suggestion
SmartSim, applications using the SmartRedis clients are directly connected to
```

File: README.md [hunk 32]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+
+The following code will show the peices of the simulation that are needed to
+transmit the data needed to plot timesteps of the simulation.
+
+```Python
+# fv_sim.py
+from smartredis import Client
+import numpy as np
+
+# initialization code ommitted
```
Comment (EricGustin) [inline README.md:668]:
```suggestion
# initialization code omitted
```

File: README.md [hunk 33]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+for i in range(0, time_steps):
+    client.poll_key(f"data_{str(i)}", 10, 1000)
+    dataset = client.get_dataset(f"data_{str(i)}")
+    ux, uy = dataset.get_tensor("ux"), dataset.get_tensor("uy")
+
+    # analysis/plotting code omitted
+
+exp.stop(db)
+```
+More details about online anaylsis with SmartSim and the full code examples can be found in the
```
Comment (EricGustin) [inline README.md:736]:
```suggestion
More details about online analysis with SmartSim and the full code examples can be found in the
```

File: README.md [hunk 34]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+# results are not retrieved immediately in case they need
+# to be fed to another function/model
+
+U = client.get_tensor("U")
+S = client.get_tensor("S")
+V = client.get_tensor("V")
+print(f"U: {U}, S: {S}, V: {V}")
+```
+
+The processing capabilties make it simple to form computational piplines of
```
Comment (EricGustin) [inline README.md:786]:
```suggestion
The processing capabilities make it simple to form computational pipelines of
```

**Review by EricGustin — CHANGES_REQUESTED** [review]
Changed most typo comments to suggestions.

File: README.md [hunk 35]
```diff
@@ -90,43 +453,435 @@ API with additions specific to scientific workflows.
...
+functions, scripts, and models.
+
+See the full [TorchScript Language Reference](https://pytorch.org/docs/stable/jit.html#torchscript-language)
+documentation for more information on available methods, functions, and how
+to create your own.
+
+## Online Inference
+
+SmartSim supports the following frameworks for quering Machine Learning models
```
Comment (EricGustin) [inline README.md:795]:
```suggestion
SmartSim supports the following frameworks for querying Machine Learning models
```

File: doc/tutorials/inference.rst [hunk 36]
```diff
@@ -0,0 +1,351 @@
...
+    client.set_model("kmeans", model, "ONNX", device="CPU")
+    client.run_model("kmeans", inputs="input", outputs=["labels", "transform"])
+
+    print(client.get_tensor("labels"))
+
+
+Random Forest
+-------------
+
+The Random Forest example uses the Iris datset from Scikit Learn to train a
```
Comment (EricGustin) [inline doc/tutorials/inference.rst:330]:
```suggestion
The Random Forest example uses the Iris dataset from Scikit Learn to train a
```

**Pull request merged**